packages/doenetml-worker-rust/lib-doenetml-core/src/core/graph_node_lookup.rs

```diff
@@ -0,0 +1,167 @@
+        self.keys().map(|node| (node, self.get_tag(&node).unwrap()))
+    }
+}
+
+impl<T> Taggable<GraphNode, T> for GraphNodeLookup<T> {
+    fn get_tag(&self, node: &GraphNode) -> Option<&T> {
+        match node {
+            GraphNode::Component(idx) => self.components.get(*idx).and_then(|x| x.as_ref()),
+            GraphNode::String(idx) => self.strings.get(*idx).and_then(|x| x.as_ref()),
+            GraphNode::Prop(idx) => self.props.get(*idx).and_then(|x| x.as_ref()),
+            GraphNode::State(idx) => self.states.get(*idx).and_then(|x| x.as_ref()),
+            GraphNode::Query(idx) => self.queries.get(*idx).and_then(|x| x.as_ref()),
+            GraphNode::Virtual(idx) => self.virtuals.get(*idx).and_then(|x| x.as_ref()),
+        }
+    }
+    fn set_tag(&mut self, node: GraphNode, tag: T) {
+        let idx = node.idx();
+        match node {
+            GraphNode::Component(_) => {
+                if idx >= self.components.len() {
+                    self.components.resize_with(idx + 1, || None);
+                }
+                self.components[idx] = Some(tag);
+            }
+            GraphNode::String(_) => {
+                if idx >= self.strings.len() {
+                    self.strings.resize_with(idx + 1, || None);
+                }
+                self.strings[idx] = Some(tag);
+            }
+            GraphNode::Prop(_) => {
+                if idx >= self.props.len() {
+                    self.props.resize_with(idx + 1, || None);
+                }
+                self.props[idx] = Some(tag);
+            }
+            GraphNode::State(_) => {
+                if idx >= self.states.len() {
+                    self.states.resize_with(idx + 1, || None);
+                }
+                self.states[idx] = Some(tag);
+            }
+            GraphNode::Query(_) => {
+                if idx >= self.queries.len() {
+                    self.queries.resize_with(idx + 1, || None);
+                }
+                self.queries[idx] = Some(tag);
+            }
+            GraphNode::Virtual(_) => {
+                if idx >= self.virtuals.len() {
+                    self.virtuals.resize_with(idx + 1, || None);
+                }
+                self.virtuals[idx] = Some(tag);
+            }
+        }
+    }
+}
+
+/// Data structure to allow for lookup up of values keyed by `(GraphNode, GraphNode)`.
+#[derive(Clone, Debug, Default)]
+pub struct DoubleNodeLookup<T> {
+    first: GraphNodeLookup<usize>,
+    second: Vec<GraphNodeLookup<T>>,
+}
+
+impl<T> DoubleNodeLookup<T> {
+    pub fn new() -> Self {
+        Self {
+            first: GraphNodeLookup::new(),
+            second: Vec::new(),
+        }
+    }
+
+    pub fn get(&self, key: &(GraphNode, GraphNode)) -> Option<&T> {
+        let (first, second) = key;
+        //let (second, first) = key;
+        let idx = self.first.get_tag(first)?;
+        // This should be safe because only indices of inserted items appear in `first`.
```

Comment:
I'm confused why this comment is needed since the following line still uses `?`. Given the comment, I was expecting the next line to be
```
self.second[*idx].get_tag(second)
```